### PR TITLE
[scudo] Add EnableMultiRegions mode

### PR DESCRIPTION
**Pull request by ChiaHungDuan**
Instead of having single region for a size class, this mode increases
    the number of regions when one is exhausted. This reduces the
    fragmentation issue when one region is exhausted and gives the finer
    grunularity of choosing the size classes. However, it inevitably
    introduces some performance overhead because of the management of
    several regions for single size class. Currently, it's an experimental
    option and expected to see performance turbulance for a little bit.
    
    See more details and constraints in allocator_config.def.

### CODE REVIEW

**Comment by ChiaHungDuan** [comment]
@cferris1000, there are two commits in this pull request. I'll split them up later. Now it's used for understanding the dependency

File: compiler-rt/lib/scudo/standalone/allocator_config.def [hunk 1]
```diff
@@ -84,6 +84,20 @@ PRIMARY_OPTIONAL(const uptr, CompactPtrScale, SCUDO_MIN_ALIGNMENT_LOG)
 // Indicates support for offsetting the start of a region by a random number of
 // pages. This is only used if `EnableContiguousRegions` is enabled.
 PRIMARY_OPTIONAL(const bool, EnableRandomOffset, false)
+
+// This allows each size class to have multiple regions instead of one. Note
+// that this is an experimental option so it has a few constraints while using.
+//   a. Pointer compaction is diabled. Which means `CompactPtrT` needs to be the
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/allocator_config.def:90]: diabled -> disabled

to be the -> to be a

File: compiler-rt/lib/scudo/standalone/allocator_config.def [hunk 2]
```diff
@@ -84,6 +84,20 @@ PRIMARY_OPTIONAL(const uptr, CompactPtrScale, SCUDO_MIN_ALIGNMENT_LOG)
...
+//      pointer integral type, i.e., uptr.
+//   b. `EnableRandomOffset` needs to be false. Pointer grouping requires
+//      the beginning of allocation address of a region to be aligned with
+//      `GroupSizeLog`. Without pointer compaction, it relies the region to be
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/allocator_config.def:94]: relies -> requires

File: compiler-rt/lib/scudo/standalone/allocator_config.def [hunk 3]
```diff
@@ -84,6 +84,20 @@ PRIMARY_OPTIONAL(const uptr, CompactPtrScale, SCUDO_MIN_ALIGNMENT_LOG)
...
+//      allocated with proper alignment and the random offset will break the
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/allocator_config.def:95]: and the random offset -> and a random offset

File: compiler-rt/lib/scudo/standalone/allocator_config.def [hunk 4]
```diff
@@ -84,6 +84,20 @@ PRIMARY_OPTIONAL(const uptr, CompactPtrScale, SCUDO_MIN_ALIGNMENT_LOG)
...
+//      assumption.
+//   c. Condition variable is not supported under this mode. This is still under
+//      developing.
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/allocator_config.def:98]: developing -> development

File: compiler-rt/lib/scudo/standalone/mem_map_linux.cpp [hunk 5]
```diff
@@ -127,11 +127,26 @@ void MemMapLinux::releaseAndZeroPagesToOSImpl(uptr From, uptr Size) {
...
+                                     uptr Flags, uptr Alignment) {
   ReservedMemoryLinux::MemMapT MemMap;
-  if (!MemMap.map(Addr, Size, Name, Flags | MAP_NOACCESS))
+  uptr MapSize = Size;
+  if (Alignment != getPageSizeCached())
+    MapSize += Alignment;
+  if (!MemMap.map(Addr, MapSize, Name, Flags | MAP_NOACCESS))
     return false;
 
+  if (Alignment != getPageSizeCached()) {
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/mem_map_linux.cpp:138]: Is it worth adding a comment indicating that this is not the normal path?

File: compiler-rt/lib/scudo/standalone/mem_map_base.h [hunk 6]
```diff
@@ -93,9 +93,11 @@ template <class Derived, typename MemMapTy> class ReservedMemory {
   constexpr ReservedMemory() = default;
 
   // Reserve a chunk of memory at a suggested address.
-  bool create(uptr Addr, uptr Size, const char *Name, uptr Flags = 0) {
+  bool create(uptr Addr, uptr Size, const char *Name, uptr Flags = 0,
+              uptr Alignment = getPageSizeCached()) {
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/mem_map_base.h:97]: Would it be better to change the alignment to be something like AlignmentPages and have the default be 1?

A lot of the places are checking that alignment is equal to getPageSizeCached() which is free on Android (most of the time), but is not as free on devices that don't have a hard-coded page size.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 7]
```diff
@@ -56,6 +56,17 @@ template <typename Config> class SizeClassAllocator64 {
   static_assert(RegionSizeLog >= GroupSizeLog,
                 "Group size shouldn't be greater than the region size");
   static const uptr GroupScale = GroupSizeLog - CompactPtrScale;
+  // Local cache stores the pointers in the type of compacted pointer and the
+  // compaction is done by calculating the offset to the base address of a
+  // region. Currently, we don't support decompacting through multiple regions
+  // because of the concern of performance and so we disable the pointer
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:62]: because of ... -> due to performance concerns, we disable pointer compaction.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 8]
```diff
@@ -56,6 +56,17 @@ template <typename Config> class SizeClassAllocator64 {
...
+  // compaction.
+  // TODO(chiahungduan): Allow local cache store the raw pointer and keep
+  // storing the compacted pointers in each region to save memory.
+  static const bool DisablePtrCompaction = Config::getEnableMultiRegions();
+  static_assert(!DisablePtrCompaction || sizeof(CompactPtrT) == sizeof(uptr),
+                "Pointer compaction is disabled, `CompactPtrT` needs to be the "
+                "same size of `uptr`");
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:69]: same size of -> same size as

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 9]
```diff
@@ -117,35 +128,45 @@ template <typename Config> class SizeClassAllocator64 {
...
     for (uptr I = 0; I < NumClasses; I++)
-      getRegionInfo(I)->RandState = getRandomU32(&Seed);
+      RegionInfoManager.getCurRegionInfo(I)->RandState = getRandomU32(&Seed);
 
     if (Config::getEnableContiguousRegions()) {
       ReservedMemoryT ReservedMemory = {};
+      // Block grouping requires the base address of a Region to be aligned
+      // with GrouopSize and pointer is compacted according to the offset to the
+      // base of a region so it always meets the requirement. As a result when
+      // the compaction is disabled, it relies the base address to be aligned.
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:146]: it relies the base address -> it relies on the base address

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 10]
```diff
@@ -155,82 +176,111 @@ template <typename Config> class SizeClassAllocator64 {
 
   void unmapTestOnly() {
     for (uptr I = 0; I < NumClasses; I++) {
-      RegionInfo *Region = getRegionInfo(I);
-      {
-        ScopedLock ML(Region->MMLock);
-        MemMapT MemMap = Region->MemMapInfo.MemMap;
+      auto RegionInfoIter = RegionInfoManager.getRegionInfoIter(I);
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:179]: Would this be better as a for loop rather than a do while loop?

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 11]
```diff
@@ -155,82 +176,111 @@ template <typename Config> class SizeClassAllocator64 {
...
-      ScopedLock FL(Region->FLLock);
-      const uptr BlockSize = getSizeByClassId(I);
-      uptr TotalBlocks = 0;
-      for (BatchGroupT &BG : Region->FreeListInfo.BlockList) {
-        // `BG::Batches` are `TransferBatches`. +1 for `BatchGroup`.
-        BatchClassUsedInFreeLists += BG.Batches.size() + 1;
-        for (const auto &It : BG.Batches)
-          TotalBlocks += It.getCount();
-      }
+      auto RegionInfoIter = RegionInfoManager.getRegionInfoIter(I);
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:202]: Same as above, this seems like it would be a bit easier to understand as a for loop.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 12]
```diff
@@ -274,8 +347,24 @@ template <typename Config> class SizeClassAllocator64 {
...
+      // Thread-safety annotation doesn't support lambda. Use a runtime check
+      // instead.
+      Region->MMLock.assertHeld();
+      const uptr RegionEnd = Region->MemMapInfo.MemMap.getBase() +
+                             Region->MemMapInfo.MemMap.getCapacity();
+      return Ptr >= Region->RegionBeg && Ptr < RegionEnd;
+    };
+
+    // When multiple-regions is enabled, we need to sort the array to dispatch
+    // the blocks to different regions efficiently. Thus even we don't put
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:361]: Thus even we -> If we

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 13]
```diff
@@ -274,8 +347,24 @@ template <typename Config> class SizeClassAllocator64 {
...
+    // BatchClass into groups, sorting is still necessary and it'll be handled
+    // later in the function.
+    // TODO: Reorder the use of variable
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:364]: Is there some more to this TODO? I don't quite understand what this is referencing.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 14]
```diff
@@ -408,15 +545,20 @@ template <typename Config> class SizeClassAllocator64 {
...
-    // approaches to ensure one page release is done.
-    if (Region->MMLock.tryLock()) {
-      uptr BytesReleased = releaseToOSMaybe(Region, ClassId, ReleaseType);
-      Region->MMLock.unlock();
-      return BytesReleased;
-    }
+    auto RegionInfoIter = RegionInfoManager.getRegionInfoIter(ClassId);
+
+    do {
+      // Note that the tryLock() may fail spuriously, given that it should
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:551]: This comment is slightly hard to follow. Maybe something like:

Note that the tryLock() can fail under certain circumstances. Since this should be a rare occurrence, there is no need to do anything to force at least one page to be released.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 15]
```diff
@@ -425,36 +567,54 @@ template <typename Config> class SizeClassAllocator64 {
...
+    if (DisablePtrCompaction)
+      return reinterpret_cast<void *>(CompactPtr);
     return reinterpret_cast<void *>(
         decompactPtrInternal(getCompactPtrBaseByClassId(ClassId), CompactPtr));
   }
 
   static BlockInfo findNearestBlock(const char *RegionInfoData,
                                     uptr Ptr) NO_THREAD_SAFETY_ANALYSIS {
+    if (Config::getEnableMultiRegions()) {
+      Printf("MultiRegions hasn't supported finding nearest block yet.\n");
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:615]: hasn't -> doesn't

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 16]
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
...
+    //   enclosing class in a potentially evaluated subexpression, the program
+    //   is ill-formed.
+    // So we have to `outline` the `Size`/`Array` into another struct `Storage`.
+    struct Storage {
+      RegionInfo *pop() {
+        if (Size == NumEntries)
+          return nullptr;
+        return &Array[Size++];
+      }
+      // The amount memory used by this allocator is about (NumEntries *
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:761]: The amount memory -> The amount of memory

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 17]
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
...
+      // RegionSize). For example, region with size 256 KB will have 2GB space
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:762]: region with size 256 KB -> a 256 KB region

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 18]
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
...
+      DCHECK_LT(ClassId, NumClasses);
+      return CurrentRegionInfo[ClassId].P;
+    }
+
+    ALWAYS_INLINE RegionInfoIter getRegionInfoIter(uptr ClassId) {
+      return RegionInfoIter(LowestAddrRegionInfo[ClassId].P,
+                            RegionInfoLock[ClassId]);
+    }
+
+    // RegionInfos for the same size class will be stored in the order of base
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:849]: store in the order of base  address -> ordered by base address

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 19]
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
...
+    // address. Which means every RegionInfo visiting will be starting from
+    // lowest address and which aligns with how pointer grouping works.
+    void pushRegionInfo(RegionInfo *Region, uptr ClassId)
+        REQUIRES(Region->MMLock) {
+      DCHECK_LT(ClassId, NumClasses);
+      DCHECK(Region->MemMapInfo.MemMap.isAllocated());
+
+      // The creation of new region requires holding the MMLock of current
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:857]: of new region -> of a new region

of current -> of the current

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 20]
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
...
+        // update the `LowestAddrRegionInfo` by copying the `CurrentRegionInfo`.
+        for (uptr I = 0; I < NumClasses; ++I)
+          CHECK_EQ(CurrentRegionInfo[I].P, LowestAddrRegionInfo[I].P);
+      }
+      scudo::shuffle(CurrentRegionInfo, NumClasses, Seed);
+      memcpy(CurrentRegionInfo, LowestAddrRegionInfo,
+             sizeof(RegionInfoPointer) * NumClasses);
+    }
+
+    // Scudo requires the data member constant initializable. Array of raw
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:906]: Array of raw -> An array of raw

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 21]
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
...
+    // pointers doesn't meet the condition. Therefore, wrap the pointer in the
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:907]: meet the condition -> meet this condition

File: compiler-rt/lib/scudo/standalone/allocator_config.def (hunk 1 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/allocator_config.def]: Done.

File: compiler-rt/lib/scudo/standalone/allocator_config.def (hunk 2 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/allocator_config.def]: Done.

File: compiler-rt/lib/scudo/standalone/allocator_config.def (hunk 3 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/allocator_config.def]: Done.

File: compiler-rt/lib/scudo/standalone/allocator_config.def (hunk 4 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/allocator_config.def]: Done.

File: compiler-rt/lib/scudo/standalone/mem_map_base.h (hunk 6 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/mem_map_base.h]: This is a good idea. Done.

BTW, I haven't rebased so I haven't changed to use `getPageSizeLogCached` accordingly. Will do it later.

File: compiler-rt/lib/scudo/standalone/mem_map_linux.cpp (hunk 5 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/mem_map_linux.cpp]: Add `LIKELY`, let me know if you think a comment is better.

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 10 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h:179]: There must be at least one RegionInfo instance (for multi-regions case, we allocate one in init()) so I take do-while loop when iterating RegionInfos. Let me know if you think it's still better for understanding (I'm fine with either way)

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 12 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Done.

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 13 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Sorry, that's some private note for myself and which has been done.

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 14 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Change the description to what you suggested and with minor change in the end. Which is
```
... there is no need to do anything to force at least `releaseToOSMaybe()` is called().
```

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 15 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Done.

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 16 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Done

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 17 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Done.

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 18 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Done.

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 19 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Done.

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 20 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Done.

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 21 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/primary64.h]: Done.

File: compiler-rt/lib/scudo/standalone/mem_map_linux.cpp [hunk 22]
```diff
@@ -127,10 +127,29 @@ void MemMapLinux::releaseAndZeroPagesToOSImpl(uptr From, uptr Size) {
...
-                                     uptr Flags) {
+                                     uptr Flags, uptr AlignmentPages) {
   ReservedMemoryLinux::MemMapT MemMap;
-  if (!MemMap.map(Addr, Size, Name, Flags | MAP_NOACCESS))
-    return false;
+  const bool NeedToAdjustAlignment = AlignmentPages != 1;
+  uptr MapSize = Size;
+
+  if (LIKELY(!NeedToAdjustAlignment)) {
+    if (!MemMap.map(Addr, MapSize, Name, Flags | MAP_NOACCESS))
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/mem_map_linux.cpp:136]: Should this have MAP_ALLOWNOMEM in the flags? Same with theMap calls below.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 23]
```diff
@@ -117,35 +127,46 @@ template <typename Config> class SizeClassAllocator64 {
...
       Seed = static_cast<u32>(Time ^ (reinterpret_cast<uptr>(&Seed) >> 12));
 
     for (uptr I = 0; I < NumClasses; I++)
-      getRegionInfo(I)->RandState = getRandomU32(&Seed);
+      RegionInfoManager.getCurRegionInfo(I)->RandState = getRandomU32(&Seed);
 
     if (Config::getEnableContiguousRegions()) {
       ReservedMemoryT ReservedMemory = {};
+      // Block grouping requires the base address of a Region to be aligned
+      // with GrouopSize and pointer is compacted according to the offset to the
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:143]: GrouopSize -> GroupSize

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 10 above)
Comment (cferris1000) [reply compiler-rt/lib/scudo/standalone/primary64.h:179]: Seems reasonable to me.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 24]
```diff
@@ -117,35 +127,46 @@ template <typename Config> class SizeClassAllocator64 {
...
+      // base of a region so it always meets the requirement. As a result when
+      // the compaction is disabled, it relies on the base address to be
+      // aligned.
+      const uptr Alignment =
+          DisablePtrCompaction ? (1UL << GroupSizeLog) : PageSize;
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:148]: Is there a way to add a DCHECK that would fail if CompactPtrT is set but pointer compaction is disabled?

File: compiler-rt/lib/scudo/standalone/primary64.h (hunk 11 above)
Comment (cferris1000) [reply compiler-rt/lib/scudo/standalone/primary64.h:202]: As above, your explanation seems reasonable.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 25]
```diff
@@ -155,82 +176,111 @@ template <typename Config> class SizeClassAllocator64 {
...
+
+      ++RegionInfoIter;
+    } while (!RegionInfoIter.end());
 
     bool ReportRegionExhausted = false;
 
-    if (conditionVariableEnabled()) {
+    RegionInfo *Region = RegionInfoManager.getCurRegionInfo(ClassId);
+
+    // TODO(chiahungduan): Support multiple-regions with condition variable.
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:282]: Similar question, can we add  DCHECK that will fail if condition is enabled but isn't supported.

Better if it could be a static_assert to cause a compilation failure.

File: compiler-rt/lib/scudo/standalone/primary64.h [hunk 26]
```diff
@@ -408,15 +544,20 @@ template <typename Config> class SizeClassAllocator64 {
...
-      uptr BytesReleased = releaseToOSMaybe(Region, ClassId, ReleaseType);
-      Region->MMLock.unlock();
-      return BytesReleased;
-    }
+    auto RegionInfoIter = RegionInfoManager.getRegionInfoIter(ClassId);
+
+    do {
+      // Note that the tryLock() can fail under certain circumstances. Since
+      // this should be a rare occurrence, there is no need to do anything to
+      // force at least one `releaseToOSMaybe()` is called().
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/primary64.h:552]: This is slightly awkward. Hope about something like:

force at least one call to `releaseToOSMaybe()`.

**Comment by cferris1000** [comment]
I've got a merged version of this that I'll be uploading sometime soon. So closing this one.
